daemon/src/surface.rs

```diff
@@ -207,12 +209,35 @@ impl Surface {
                     bail!("Directory {path:?} is empty");
                 }
 
-                let img_path = files[rand::random::<usize>() % files.len()].clone();
+                // Reset our index if we'd be out of bounds
+                if self.idx >= files.len() {
+                    self.idx = 0
+                };
+
+                // This is a bit unweildy. We're sorting every time thru this loop. But since we're
+                // walking the directory each time, I guess it's not that bad. Sorting can't be slower
+                // than reading paths off the disk, right?
+                let img_path = if self.wallpaper_info.sorting.eq(&Some(String::from("natural"))) {
+                    // You can't sort an immutable Vec, so we have to clone it as mutable.
+                    let mut sorted: Vec<PathBuf> = files.clone();
+                    sorted.sort();
+                    sorted[self.idx].clone()
+                } else if self.wallpaper_info.sorting.eq(&Some(String::from("reverse"))) {
+                    // You can't sort an immutable Vec, so we have to clone it as mutable.
+                    let mut reversed: Vec<PathBuf> = files.clone();
+                    reversed.sort_by(|a,b| b.cmp(a));
+                    reversed[self.idx].clone()
+                } else {
+                    files[rand::random::<usize>() % files.len()].clone()
+                };
```

Comment:
If you go with an `enum`, you'll probably want to switch over to:

```rust
let img_path = match self.wallpaper_info.sorting {
    Sorting::Random => { /* … */ },
    Sorting::Natural => { /* … */ },
    Sorting::Reverse => { /* … */ },
};
```

I note also that `files` isn't used after this block, which means you can move `files` rather than cloning it:

```rust
let mut sorted = files;
sorted.sort();
```

Likewise, since you need a `PathBuf` and the `Vec<PathBuf>` isn't used after picking one, you can move the selected `PathBuf` out of the `Vec` rather than cloning your selection:

```rust
let img_path = match sorting {
  /* … */
  Sorting::Random => {
    let index = rand::random::<usize>() % files.len();
    files.into_iter().nth(index).next().unwrap()
  }
};
```

Reply:
Again, yes, yes, yes! 

Learning more about copying and borrowing vs cloning things. Thanks for the tip!